### DISCUSSION TITLE
How TTF is rendered?

### DISCUSSION THREAD

**Discussion by dmitrykolesnikovich** [bevyengine/bevy#16969] [Q&A] [answered]
From [source code](https://github.com/bevyengine/bevy/tree/main/crates/bevy_text/src) it's not clear for me how TTF is actually parsed and then rendered. 
- Do you actually parse TTF file using OpenType spec or do you use external library for parsing? 
- Do you use bitmap atlas font to actually draw glyphs on the screen and if yes how do you even make it not blurry when letters are scaled too large or too small?

**Comment by rparrett** [comment, marked as answer]
> Do you actually parse TTF file using OpenType spec or do you use external library for parsing?

We use https://github.com/pop-os/cosmic-text, which uses https://github.com/harfbuzz/ttf-parser.

> Do you use bitmap atlas font to actually draw glyphs on the screen and if yes how do you even make it not blurry when letters are scaled too large or too small?

Yes, glyphs are rasterized into an atlas. The resolution of the rasters is based on the `font_size` you specify. It's sort of up to you to use an appropriate value for `font_size` based on your projection / scale.

If [font smoothing](https://docs.rs/bevy/latest/bevy/prelude/struct.TextFont.html#structfield.font_smoothing) is disabled, the glyph atlases will use nearest neighbor sampling. (This also applies a 50% opacity threshold to the glyphs).
